File: scripts/verify-public-mirror.mjs
/**
 * Ensures files changed by the current frontend commit also exist with the
 * same contents in the public monorepo checkout.  CI checks out that repo at
 * ../chatty-public before invoking this script.
 */
import { execFileSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import path from "node:path";

const mirrorRoot = process.env.PUBLIC_MIRROR_DIR;
if (!mirrorRoot) throw new Error("PUBLIC_MIRROR_DIR must point to a PersonaliAI/chatty checkout.");

const changed = execFileSync("git", ["diff", "--name-only", "HEAD^", "HEAD"], { encoding: "utf8" })
  .split(/\r?\n/)
  .filter(Boolean)
  .filter((file) => file.startsWith("src/") || file.startsWith("public/") || ["package.json", ".env.example"].includes(file));

const failures = [];
for (const file of changed) {
  const mirrorFile = path.join(mirrorRoot, "frontend", file);
  if (!existsSync(mirrorFile)) {
    failures.push(`${file}: missing from public mirror`);
    continue;
  }
  if (!readFileSync(file).equals(readFileSync(mirrorFile))) failures.push(`${file}: contents differ from public mirror`);
}

if (failures.length) {
  console.error("Public mirror verification failed:\n" + failures.map((failure) => `  - ${failure}`).join("\n"));
  process.exit(1);
}
console.log(`Public mirror verified for ${changed.length} changed frontend file(s).`);

File: backend/scripts/verify-public-mirror.py
"""Verify changed backend files are present in the public monorepo mirror."""

from __future__ import annotations

import os
import subprocess
import sys
from pathlib import Path

MIRRORED_PREFIXES = (
    "main.py",
    "app/",
    "plugins/",
    "tests/",
    "sql/",
    "scripts/",
    "requirements.txt",
    "Dockerfile",
    "voice-agent/",
    "env.yaml",
    "env.yaml.example",
    "supabase/migrations/",
)


def is_mirrored(path: str) -> bool:
    return any(path == prefix or path.startswith(prefix) for prefix in MIRRORED_PREFIXES)


def main() -> int:
    mirror_root = os.environ.get("PUBLIC_MIRROR_DIR")
    if not mirror_root:
        raise RuntimeError("PUBLIC_MIRROR_DIR must point to a PersonaliAI/chatty checkout.")

    changed = subprocess.check_output(
        ["git", "diff", "--name-only", "HEAD^", "HEAD"],
        text=True,
    ).splitlines()

    failures: list[str] = []
    for rel in filter(is_mirrored, changed):
        private_file = Path(rel)
        public_file = Path(mirror_root) / "backend" / rel
        if not public_file.exists():
            failures.append(f"{rel}: missing from public mirror")
            continue
        if private_file.read_bytes() != public_file.read_bytes():
            failures.append(f"{rel}: contents differ from public mirror")

    if failures:
        print("Public backend mirror verification failed:", file=sys.stderr)
        for failure in failures:
            print(f"  - {failure}", file=sys.stderr)
        return 1

    print(f"Public backend mirror verified for {sum(1 for rel in changed if is_mirrored(rel))} changed file(s).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
